src/esp/assets/GenericInstanceMeshData.cpp

```diff
@@ -229,17 +255,24 @@ void GenericInstanceMeshData::uploadBuffersToGPU(bool forceReload) {
   renderingBuffer_ =
       std::make_unique<GenericInstanceMeshData::RenderingBuffer>();
 
-  /* Pack primitive IDs into a texture. 1D texture won't be large enough so the
-     data have to be put into a 2D texture. For simplicity on both the C++ and
-     shader side the texture has a fixed width and height is dynamic, and
-     addressing is done as (gl_PrimitiveID % width, gl_PrimitiveID / width). */
-  ASSERT(objectIds_.size() == cpu_ibo_.size());
-  renderingBuffer_->tex = createInstanceTexture(objectIds_);
+  std::vector<Mn::UnsignedInt> objectIdIndices = removeDuplicates(objectIds_);
+  std::vector<Mn::UnsignedInt> indices = Mn::MeshTools::combineIndexedArrays(
+      std::make_pair(std::cref(objectIdIndices), std::ref(objectIds_)),
+      std::make_pair(std::cref(cpu_ibo_), std::ref(cpu_vbo_)),
+      std::make_pair(std::cref(cpu_ibo_), std::ref(cpu_cbo_)));
+  renderingBuffer_->obo.setData(objectIds_,
+                                Magnum::GL::BufferUsage::StaticDraw);
   renderingBuffer_->vbo.setData(cpu_vbo_, Magnum::GL::BufferUsage::StaticDraw);
   renderingBuffer_->cbo.setData(cpu_cbo_, Magnum::GL::BufferUsage::StaticDraw);
-  renderingBuffer_->ibo.setData(cpu_ibo_, Magnum::GL::BufferUsage::StaticDraw);
+  renderingBuffer_->ibo.setTargetHint(
+      Magnum::GL::Buffer::TargetHint::ElementArray);
+  renderingBuffer_->ibo.setData(indices, Magnum::GL::BufferUsage::StaticDraw);
   renderingBuffer_->mesh.setPrimitive(Magnum::GL::MeshPrimitive::Triangles)
-      .setCount(cpu_ibo_.size() * 3)
+      .setCount(cpu_ibo_.size())
+      .addVertexBuffer(renderingBuffer_->obo, 0,
+                       gfx::PrimitiveIDTexturedShader::ObjectId{
+                           gfx::PrimitiveIDTexturedShader::ObjectId::DataType::
+                               UnsignedShort})
       .addVertexBuffer(renderingBuffer_->vbo, 0,
```

Comment:
Now that we have three buffers, I think the `obo`, `vbo`, `cbo` could be interleaved into a single buffer (you could use [MeshTools::interleave()](https://doc.magnum.graphics/magnum/namespaceMagnum_1_1MeshTools.html#a006bda8a1901669404adcc50b4b68735)) -- it'll give the GPU a better memory locality in the vertex shader and the setup would be just a single `addVertexBuffer()` call:

```cpp
// order of arguments to interleave() and to addVertexBuffer() has to match ;)
renderingBuffer_->vbo.setData(MeshTools::interleave(cpu_vbo_, cpu_cbo_, cpu_obo_); 
renderingBuffer_->mesh.addVertexBuffer(renderingBuffer_->vbo, 0,
    gfx::PrimitiveIDTexturedShader::Position{},
    gfx::PrimitiveIDTexturedShader::Color3{
              gfx::PrimitiveIDTexturedShader::Color3::DataType::UnsignedByte,
              gfx::PrimitiveIDTexturedShader::Color3::DataOption::Normalized}),
    gfx::PrimitiveIDTexturedShader::ObjectId{
               gfx::PrimitiveIDTexturedShader::ObjectId::DataType::UnsignedShort});
```

For an even more minimalistic setup, you could `std::move(vbo)` into the `addVertexBuffer()` call (and the same for `setIndexBuffer()`), then you need to store just the mesh and all buffers are owned by it ([docs on this](https://doc.magnum.graphics/magnum/classMagnum_1_1GL_1_1Mesh.html#GL-Mesh-buffer-ownership)).

Reply:
I will try these. I may punt some of this to a separate change if it gets too involved.

Reply:
I tried interleave but it didn't work right away. Will debug it later and submit as a follow-up patch.

```
diff --git a/src/esp/assets/GenericInstanceMeshData.cpp b/src/esp/assets/GenericInstanceMeshData.cpp
index 931f218..58329b5 100644
--- a/src/esp/assets/GenericInstanceMeshData.cpp
+++ b/src/esp/assets/GenericInstanceMeshData.cpp
@@ -11,6 +11,7 @@
 #include <Magnum/ImageView.h>
 #include <Magnum/Math/Functions.h>
 #include <Magnum/MeshTools/CombineIndexedArrays.h>
+#include <Magnum/MeshTools/Interleave.h>
 #include <Magnum/MeshTools/RemoveDuplicates.h>
 #include <Magnum/PixelFormat.h>
 #include <Magnum/Trade/Trade.h>
@@ -241,26 +242,21 @@ void GenericInstanceMeshData::uploadBuffersToGPU(bool forceReload) {
       std::make_pair(std::cref(cpu_ibo_), std::ref(cpu_vbo_)),
       std::make_pair(std::cref(cpu_ibo_), std::ref(cpu_cbo_)));
   LOG(INFO) << "Mandeep: after: " << cpu_vbo_.size();
-  renderingBuffer_->obo.setData(objectIds_,
-                                Magnum::GL::BufferUsage::StaticDraw);
-  renderingBuffer_->vbo.setData(cpu_vbo_, Magnum::GL::BufferUsage::StaticDraw);
-  renderingBuffer_->cbo.setData(cpu_cbo_, Magnum::GL::BufferUsage::StaticDraw);
+  renderingBuffer_->vbo.setData(
+      Mn::MeshTools::interleave(cpu_vbo_, cpu_cbo_, objectIds_),
+      Magnum::GL::BufferUsage::StaticDraw);
   renderingBuffer_->ibo.setTargetHint(
       Magnum::GL::Buffer::TargetHint::ElementArray);
   renderingBuffer_->ibo.setData(indices, Magnum::GL::BufferUsage::StaticDraw);
   renderingBuffer_->mesh.setPrimitive(Magnum::GL::MeshPrimitive::Triangles)
       .setCount(cpu_ibo_.size())
-      .addVertexBuffer(renderingBuffer_->obo, 0,
-                       gfx::PrimitiveIDTexturedShader::ObjectId{
-                           gfx::PrimitiveIDTexturedShader::ObjectId::DataType::
-                               UnsignedShort})
       .addVertexBuffer(renderingBuffer_->vbo, 0,
-                       gfx::PrimitiveIDTexturedShader::Position{})
-      .addVertexBuffer(
-          renderingBuffer_->cbo, 0,
-          gfx::PrimitiveIDTexturedShader::Color3{
-              gfx::PrimitiveIDTexturedShader::Color3::DataType::UnsignedByte,
-              gfx::PrimitiveIDTexturedShader::Color3::DataOption::Normalized})
+                       gfx::PrimitiveIDTexturedShader::Position{},
+                       gfx::PrimitiveIDTexturedShader::Color3{
+                           gfx::PrimitiveIDTexturedShader::Color3::DataType::UnsignedByte,
+                           gfx::PrimitiveIDTexturedShader::Color3::DataOption::Normalized},
+                       gfx::PrimitiveIDTexturedShader::ObjectId{
+                            gfx::PrimitiveIDTexturedShader::ObjectId::DataType::UnsignedShort})
       .setIndexBuffer(renderingBuffer_->ibo, 0,
                       Magnum::GL::MeshIndexType::UnsignedInt);

```

Reply:
Hmm. Don't see anything wrong with the diff. How exactly it didn't work? Black screen of death?

Maybe the driver is unhappy about data alignment there? If I'm counting correctly, each vertex is 17 bytes (12 bytes for position, 3 for color, 2 for object ID). Can you try adding 3 byte padding after the color to make it 20 bytes (and so the floats are always at 4-byte-aligned positions and object ID at 2-byte-aligned)? It's done by specifying a number in both `interleave()` and `addVertexBuffer()`:

```cpp
renderingBuffer_->vbo.setData(MeshTools::interleave(cpu_vbo_, cpu_cbo_, 3, objectIds_)); 
renderingBuffer_->mesh.addVertexBuffer(renderingBuffer_->vbo, 0,
    gfx::PrimitiveIDTexturedShader::Position{},
    gfx::PrimitiveIDTexturedShader::Color3{
          gfx::PrimitiveIDTexturedShader::Color3::DataType::UnsignedByte,
          gfx::PrimitiveIDTexturedShader::Color3::DataOption::Normalized}),
    3,
    gfx::PrimitiveIDTexturedShader::ObjectId{
           gfx::PrimitiveIDTexturedShader::ObjectId::DataType::UnsignedShort});
```

Of course now, in case this makes it work, this begs a question if this interleaving really helps perf (the memory locality is better, but at the same time, each vertex loads 3 wasted bytes). I guess we won't know until we benchmark :)

Reply:
Yep. That was it. I forgot about alignment. Thanks!